man/hfst-info.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-INFO "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-info \- =show or test HFST versions and features
.SH SYNOPSIS
.B hfst-info
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
show or test HFST versions and features
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Test features:"
.TP
\fB\-a\fR, \fB\-\-atleast\-version\fR=\fI\,MVER\/\fR
require at least MVER version of HFST
.TP
\fB\-e\fR, \fB\-\-exact\-version\fR=\fI\,EVER\/\fR
require exactly EVER version of HFST
.TP
\fB\-m\fR, \fB\-\-max\-version\fR=\fI\,UVER\/\fR
require at most UVER version of HFST
.TP
\fB\-f\fR, \fB\-\-requirefeature\fR=\fI\,FEAT\/\fR
require named FEAT support from HFST
.PP
MVER, EVER or UVER version vectors must be composed of one to three full stop separated runs of digits.
FEAT should be name of feature supported by HFST, such as SFST, foma or openfst
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-info home page:
<https://github.com/hfst/hfst/wiki/HfstInfo>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-info.cc
//! @file hfst-format.cc
//!
//! @brief Format checking command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif


#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "hfst.hpp"

#include <set>
#include <string>

using std::set;
using std::string;

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "inc/globals-common.h"


static long min_version = -1L;
static long exact_version = -1L;
static long max_version = -1L;
static set<string> required_features;

static
long
parse_version_string(const char* s)
  {
    char* endptr;
    long major = strtoul(s, &endptr, 10L);
    if (*endptr == '\0')
      {
        return (major * 10000 * 10000);
      }
    else if (*endptr != '.')
      {
        error(EXIT_FAILURE, 0, "cannot parse version string from %s", endptr);
      }
    s = endptr + 1;
    long minor = strtoul(s, &endptr, 10L);
    if (*endptr == '\0')
      {
        return (major * 10000 * 10000) + (minor * 10000);
      }
    else if (*endptr != '.')
      {
        error(EXIT_FAILURE, 0, "cannot parse version string from %s", endptr);
      }
    s = endptr + 1;
    long patch = strtoul(s, &endptr, 10L);
    if (*endptr == '\0')
      {
        return (major * 10000 * 10000) + (minor * 10000) + patch;
      }
    else
      {
        error(EXIT_FAILURE, 0, "cannot parse version string from %s", endptr);
      }
    return -1L;
  }

void
print_usage()
{
        // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
           "show or test HFST versions and features\n"
            "\n", program_name);

    print_common_program_options(message_out);
    fprintf(message_out, "Test features:\n"
            "  -a, --atleast-version=MVER   require at least MVER version "
            "of HFST\n"
            "  -e, --exact-version=EVER     require exactly EVER version "
            "of HFST\n"
            "  -m, --max-version=UVER       require at most UVER version "
            "of HFST\n"
            "  -f, --requirefeature=FEAT    require named FEAT support "
            "from HFST\n");
    fprintf(message_out, "\n");
    fprintf(message_out, "MVER, EVER or UVER version vectors must be composed "
            "of one to three full stop separated runs of digits.\n"
            "FEAT should be name of feature supported by HFST, such as "
            "SFST, foma or openfst\n"
            "\n");
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}


int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          {"atleast-version", required_argument, 0, 'a'},
          {"exact-version", required_argument, 0, 'e'},
          {"max-version", required_argument, 0, 'm'},
          {"require-feature", required_argument, 0, 'f'},
          {0,0,0,0}
        };
        int option_index = 0;
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             "a:e:f:m:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }
        switch (c)
        {
        case 'a':
          min_version = parse_version_string(optarg);
          break;
        case 'e':
          exact_version = parse_version_string(optarg);
          break;
        case 'm':
          max_version = parse_version_string(optarg);
          break;
        case 'f':
          required_features.insert(optarg);
          break;
        case 'h':
          print_usage();
          return EXIT_SUCCESS;
        case 'V':
          print_version();
          return EXIT_SUCCESS;
        }
    }
    if ((min_version == -1L) && (max_version == -1L) && (exact_version == -1L)
        && (required_features.size() == 0) && (verbose == false))
      {
        verbose = true;
        verbose_printf("No tests selected; printing known data\n");
      }
    return EXIT_CONTINUE;
}


int main (int argc, char * argv[])
{
#ifdef WINDOWS
  _setmode(0, _O_BINARY);
#endif

  hfst_set_program_name(argv[0], "0.1", "HfstInfo");
  parse_options(argc, argv);
  if (min_version != -1L)
    {
      verbose_printf("Requiring current version %ld to be greater than %ld\n",
                     HFST_LONGVERSION, min_version);
      if (HFST_LONGVERSION < min_version)
        {
          error(EXIT_FAILURE, 0, "Version requirements not met");
        }
    }
  if (exact_version != -1L)
    {
      verbose_printf("Requiring current version %ld to be exactly %ld\n",
                     HFST_LONGVERSION, exact_version);
      if (HFST_LONGVERSION != exact_version)
        {
          error(EXIT_FAILURE, 0, "Version requirements not met");
        }
    }
  if (max_version != -1L)
    {
      verbose_printf("Requiring current version %ld to be greater than %ld\n",
                     HFST_LONGVERSION, max_version);
      if (HFST_LONGVERSION < max_version)
        {
          error(EXIT_FAILURE, 0, "Version requirements not met");
        }
    }
  for (set<string>::const_iterator f = required_features.begin();
       f != required_features.end();
       ++f)
    {
      if ((*f == "sfst") || (*f == "SFST") || (*f == "HAVE_SFST"))
        {
          verbose_printf("Requiring SFST support from library");
#ifndef HAVE_SFST
          error(EXIT_FAILURE, 0, "Required SFST support not present");
#endif
        }
      else if ((*f == "foma") || (*f == "FOMA") || (*f == "HAVE_FOMA"))
        {
          verbose_printf("Requiring foma support from library");
#ifndef HAVE_FOMA
          error(EXIT_FAILURE, 0, "Required foma support not present");
#endif
        }
      else if ((*f == "xfsm") || (*f == "XFSM") || (*f == "HAVE_XFSM"))
        {
          verbose_printf("Requiring xfsm support from library");
#ifndef HAVE_XFSM
          error(EXIT_FAILURE, 0, "Required xfsm support not present");
#endif
        }
      else if ((*f == "openfst") || (*f == "OPENFST") || (*f == "HAVE_OPENFST"))
        {
          verbose_printf("Requiring OpenFst support from library");
#ifndef HAVE_OPENFST
          error(EXIT_FAILURE, 0, "Required OpenFst support not present");
#endif
        }
      else if ((*f == "glib") || (*f == "USE_GLIB_UNICODE"))
        {
          verbose_printf("Requiring Unicode parsed by Glib");
#ifndef USE_GLIB_UNICODE
          error(EXIT_FAILURE, 0,
                "Required GLIB-based Unicode handling not presesnt");
#endif
        }
      else
        {
          error(EXIT_FAILURE, 0, "Required %s support is unrecognised "
                "and therefore assumed to be missing", f->c_str());
        }
    }
  verbose_printf("HFST info version: %s\n"
                 "HFST packaging: %s\n"
                 "HFST version: %s\n"
                 "HFST long version: %ld\n",
                 //          "HFST configuration revision: %s\n",
                 hfst_tool_version,
                 PACKAGE_STRING,
                 PACKAGE_VERSION,
                 HFST_LONGVERSION);
  //                 HFST_REVISION);
#if HAVE_OPENFST
  verbose_printf("OpenFst supported\n");
#endif
#if HAVE_SFST
  verbose_printf("SFST supported\n");
#endif
#if HAVE_FOMA
  verbose_printf("foma supported\n");
#endif
#if HAVE_XFSM
  verbose_printf("xfsm supported\n");
#endif
#if USE_GLIB_UNICODE
  verbose_printf("Unicode support: glib\n");
#elif USE_ICU_UNICODE
  verbose_printf("Unicode support: ICU\n");
#else
  verbose_printf("Unicode support: no (hfst)\n");
#endif

  return EXIT_SUCCESS;
}
